lib/spark.js

```diff
@@ -71,11 +72,25 @@ export default class Spark {
     }
 
     console.log(`Querying IPNI to find retrieval providers for ${retrieval.cid}`)
-    const { indexerResult, provider } = await queryTheIndex(retrieval.cid, stats.providerId)
+    const { indexerResult, provider, providers } = await queryTheIndex(
+      retrieval.cid,
+      stats.providerId,
+    )
     stats.indexerResult = indexerResult
 
     const providerFound = indexerResult === 'OK' || indexerResult === 'HTTP_NOT_ADVERTISED'
-    if (!providerFound) return
+    const noValidAdvertisement = indexerResult === 'NO_VALID_ADVERTISEMENT'
+
+    // In case index lookup failed due to network error or CID not found,
+    // we will not perform any retrieval
+    if (!providerFound && !noValidAdvertisement) return
+
+    // In case we fail to find a valid advertisement for the provider
+    // we will try to perform network wide retrieval from other providers
+    if (noValidAdvertisement) {
+      console.log('No valid advertisement found. Performing network-wide retrieval check...')
+      return await this.testNetworkRetrieval(providers, retrieval.cid, stats)
```

Comment:
The more I work in this codebase, the more I regret the design pattern using a mutable `stats` parameter passed around. I would like to eventually refactor the code so that every function returns a `stats` object with the relevant subset of fields.

Would you mind applying that design for this new function?

```js
  stats.networkRetrieval = await this.testNetworkRetrieval(providers, retrieval.cid, stats)
```

Also, let's find a different name for `stats.networkRetrieval`, e.g. `stats.altProvider` or `stats.altProviderCheck`

Reply:
I have changed it to `alternativeProviderCheck`.